tools/kraken.py

```diff
@@ -55,51 +55,67 @@ def build(self, db, options=None, option_string=None):
         self.execute('kraken-build', db, db, options=options,
                      option_string=option_string)
 
-    def pipeline(self, inBam, db, outReport=None, outReads=None, filterThreshold=None, numThreads=None):
-
-        with util.file.fifo(2) as (fastq1_pipe, fastq2_pipe):
-            # do not convert this to samtools bam2fq unless we can figure out how to replicate
-            # the clipping functionality of Picard SamToFastq
-            picard = tools.picard.SamToFastqTool()
-            picard_opts = {
-                'CLIPPING_ATTRIBUTE': tools.picard.SamToFastqTool.illumina_clipping_attribute,
-                'CLIPPING_ACTION': 'X'
-            }
-            picard.execute(inBam, fastq1_pipe, fastq2_pipe,
-                           picardOptions=tools.picard.PicardTools.dict_to_picard_opts(picard_opts),
-                           JVMmemory=picard.jvmMemDefault,
-                           background=True)
-
-            opts = {
-                '--threads': util.misc.sanitize_thread_count(numThreads),
-            }
+    def pipeline(self, db, inBams, outReports=None, outReads=None, filterThreshold=None, numThreads=None):
+        assert outReads is not None or outReports is not None
+
+        n_bams = len(inBams)
+        # 2n for paired fastq, 1n for kraken output
+        n_pipes = n_bams * 3
+
+        threads = util.misc.sanitize_thread_count(numThreads)
+
+        with util.file.fifo(n_pipes) as pipes:
+            fastq_pipes = pipes[:n_bams * 2]
+            kraken_output_pipes = pipes[n_bams * 2:]
 
             kraken_bin = os.path.join(self.libexec, 'kraken')
-            cmd = '''export KRAKEN_DEFAULT_DB={kraken_db}; {{ {kraken} --paired --fastq-input --threads {threads} {fastq1} {fastq2} 2>&1 1>&3 3>&- | sed '/Processed [0-9]* sequences/d'; }} \
-            3>&1 1>&2'''.format(
-                kraken_db=db,
+            env = os.environ.copy()
+            env['KRAKEN_DEFAULT_DB'] = db
+            cmd = '''{kraken} --lock-memory --paired --fastq-input --threads {threads} {outputs} {fastqs}'''.format(
```

Comment:
If `kraken` errors with an unhelpful message related to `--lock-memory`, you might want to put a message about ulimits (especially if this happens even if `kraken_execution` is `single` by default).

Reply:
Yeah I put it in the kraken fork source code. Most systems the hard + soft limit are 64 pages of 4kb. It requires changing a config file to set it to be `unlimited` and reboot. This limit is also consistent on broad cluster and AWS, so I might make the lock optional rather than default.